File: examples/dekita/.claude/hooks/issue_scope_check.py
#!/usr/bin/env python3
"""Issue編集時のスコープ確認を強制する。

Why:
    1つのIssueに異なるタスクを追加すると追跡性が低下する。
    1Issue1タスクの原則を強制することで、Issueの管理性を向上させる。

What:
    - gh issue edit --bodyコマンドを検出
    - チェックボックスのみの変更は許可（進捗更新のため）
    - 内容追加時はスコープ確認を強制しブロック
    - SKIP_ISSUE_SCOPE_CHECK環境変数でバイパス可能

Remarks:
    - ブロック型フック（内容追加時はブロック）
    - PreToolUse:Bashで発火（gh issue editコマンド）
    - issue-multi-problem-check.pyはIssue作成時のみ対象（責務分離）
    - forkセッションではSKIP環境変数を許可しない

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#2423: チェックボックス更新を許可する機能を追加
    - silenvx/dekita#2431: SKIP環境変数サポートと拒否メッセージの改善
"""

import json
import os
import re
import subprocess
import sys

from lib.execution import log_hook_execution
from lib.results import make_approve_result, make_block_result
from lib.session import is_fork_session, parse_hook_input
from lib.strings import extract_inline_skip_env, is_skip_env_enabled

HOOK_NAME = "issue-scope-check"
SKIP_ENV_NAME = "SKIP_ISSUE_SCOPE_CHECK"


def extract_issue_number(command: str) -> str | None:
    """コマンドからIssue番号を抽出する."""
    # gh issue edit 123 --body "..." or gh issue edit #123 -b "..."
    match = re.search(r"gh\s+issue\s+edit\s+#?(\d+)", command)
    if match:
        return match.group(1)
    return None


def get_current_issue_body(issue_number: str) -> str | None:
    """GitHub APIで現在のIssue bodyを取得する."""
    try:
        result = subprocess.run(
            ["gh", "issue", "view", issue_number, "--json", "body", "--jq", ".body"],
            capture_output=True,
            text=True,
            timeout=10,
            check=False,
        )
        if result.returncode == 0:
            return result.stdout.strip()
    except (subprocess.TimeoutExpired, FileNotFoundError):
        # gh CLI未インストールまたはタイムアウト時は取得失敗として扱う
        # → Noneを返し、チェックボックス判定をスキップしてブロック（安全側）
        pass
    return None


def extract_body_from_command(command: str) -> str | None:
    """コマンドから--bodyオプションの値を抽出する.

    サポートするパターン:
    1. 単純なクォート: --body "content" または --body 'content'
    2. heredoc: --body "$(cat <<'EOF' ... EOF)" 等のバリエーション
       - EOF のクォート: 'EOF', "EOF", EOF（クォートなし）
       - 終端パターン: EOF)", EOF"), EOF), EOF", EOF + 空白, EOF + 文字列末尾

    heredocパターンを先にチェックすることで、heredoc内のクォートで
    誤って終端しないようにする。
    """
    # --body "$(cat <<'EOF' ... EOF)" パターン（先にチェック）
    # heredoc内の内容を抽出する。終端は EOF)", EOF"), EOF), EOF" などを許容する
    # heredocパターンを先にチェックすることで、内部のクォートで誤マッチしない
    match = re.search(
        r"--body\s+\"\$\(\s*cat\s+<<['\"]?EOF['\"]?\s*\n"
        r"(?P<body>.*?)"
        r"\nEOF(?:\)\"|\"\)|\)|\"|\s|$)",
        command,
        re.DOTALL,
    )
    if match:
        return match.group("body")

    # クォート種別をキャプチャして、対応する終端クォートまでを非貪欲に取得する
    # heredocパターンにマッチしなかった場合のフォールバック
    match = re.search(
        r"--body\s+(?P<quote>['\"])(?P<body>.*?)(?P=quote)",
        command,
        re.DOTALL,
    )
    if match:
        return match.group("body")

    return None


def is_checkbox_only_change(old_body: str, new_body: str) -> bool:
    """変更がチェックボックスのステータス変更のみかどうかを判定する.

    チェックボックスのステータス変更:
    - [ ] → [x] または [X]
    - [x] または [X] → [ ]

    上記以外の変更がある場合はFalseを返す。

    注意:
    - old_body / new_body が None の場合は判定不能として False を返す。
    - old_body / new_body が空文字列 "" の場合も、チェックボックス変更とはみなさず False を返す。
    """
    # None や空文字列の body は「チェックボックスのみの変更」とはみなさない
    if old_body is None or new_body is None:
        return False
    if old_body == "" or new_body == "":
        return False

    # 行ごとに比較
    old_lines = old_body.splitlines()
    new_lines = new_body.splitlines()

    # 行数が異なる場合は内容追加/削除あり
    if len(old_lines) != len(new_lines):
        return False

    # Markdownでは -, *, + がリストマーカーとして有効なため、いずれも許可する
    checkbox_pattern = re.compile(r"^(\s*[-*+]\s*)\[([ xX])\](.*)$")

    for old_line, new_line in zip(old_lines, new_lines, strict=True):
        if old_line == new_line:
            continue

        # 両方がチェックボックス行かどうか
        old_match = checkbox_pattern.match(old_line)
        new_match = checkbox_pattern.match(new_line)

        if old_match and new_match:
            # プレフィックスと内容が同じで、チェック状態のみ異なる場合はOK
            if old_match.group(1) == new_match.group(1) and old_match.group(3) == new_match.group(
                3
            ):
                # チェック状態のみ異なる
                continue

        # チェックボックス以外の変更がある
        return False

    return True


def main():
    """PreToolUse hook for Bash commands."""
    try:
        data = parse_hook_input()
        tool_input = data.get("tool_input", {})
        command = tool_input.get("command", "")

        # gh issue edit コマンドを検出
        if "gh issue edit" not in command:
            # 対象外のコマンドは何も出力しない
            sys.exit(0)

        # --body オプションで内容を変更しようとしている場合のみ
        if "--body" not in command:
            # --body なしは対象外
            sys.exit(0)

        # Issue #2458: forkセッション判定（SKIPチェックの前に実行）
        session_id = data.get("session_id", "")
        source = data.get("source", "")
        transcript_path = data.get("transcript_path")
        is_fork = is_fork_session(session_id, source, transcript_path)

        # Issue #2431: SKIP環境変数のチェック（エクスポートとインライン両対応）
        # Issue #2458: forkセッションではSKIPを許可しない
        skip_requested = is_skip_env_enabled(os.environ.get(SKIP_ENV_NAME)) or is_skip_env_enabled(
            extract_inline_skip_env(command, SKIP_ENV_NAME)
        )

        if skip_requested:
            if is_fork:
                # forkセッションではSKIPを許可しない
                log_hook_execution(
                    HOOK_NAME,
                    "block",
                    f"fork-session: {SKIP_ENV_NAME} not allowed",
                )
                result = make_block_result(
                    HOOK_NAME,
                    f"""[issue-scope-check] 🚫 forkセッションではSKIP不可

forkセッションでは{SKIP_ENV_NAME}は使用できません。
forkセッションは別タスクとして扱うべきです。

【対応方法】
新しいIssueを作成してください:
gh issue create --title "..." --body "..."
""",
                )
                print(json.dumps(result))
                sys.exit(0)

            # 通常セッションではSKIPを許可
            log_hook_execution(
                HOOK_NAME,
                "approve",
                f"{SKIP_ENV_NAME}=1: スコープ確認をスキップ",
            )
            result = make_approve_result(HOOK_NAME, f"{SKIP_ENV_NAME}=1")
            print(json.dumps(result))
            sys.exit(0)

        # Issue #2423: チェックボックス更新のみの場合は許可
        issue_number = extract_issue_number(command)
        if issue_number:
            current_body = get_current_issue_body(issue_number)
            new_body = extract_body_from_command(command)

            if current_body and new_body and is_checkbox_only_change(current_body, new_body):
                # チェックボックス更新のみなので許可
                result = make_approve_result(HOOK_NAME, "checkbox status change only")
                log_hook_execution(HOOK_NAME, "approve", "checkbox status change only")
                print(json.dumps(result))
                sys.exit(0)

            # チェックボックス判定を行えなかった理由をログに残す
            if not current_body:
                log_hook_execution(HOOK_NAME, "skip", "Failed to get current issue body")
            if not new_body:
                log_hook_execution(HOOK_NAME, "skip", "Failed to extract new body from command")

        # スコープ確認を強制（ブロック）
        # Issue番号があればスキップコマンドに含める
        issue_num_for_msg = issue_number if issue_number else "<Issue番号>"
        block_message = f"""🚫 Issue編集時のスコープ確認

Issueに内容を追加する前に確認が必要です:
- 追加しようとしている内容は、元のIssueと同じタスクですか？
- 異なるタスクであれば、別のIssueとして作成すべきです
- 1 Issue = 1 タスク の原則を守ってください

【対応方法】
1. 同じタスクの場合: ユーザーに確認してから編集を続行
2. 異なるタスクの場合: gh issue create --title "..." --body "..." で新規作成

【スキップ方法】（ユーザー確認済みの場合）
```
SKIP_ISSUE_SCOPE_CHECK=1 gh issue edit {issue_num_for_msg} --body "..."
```

【補足】
- チェックボックスのステータス変更のみの場合は自動許可されます
- 行数が変わる変更（セクション追加など）はブロックされます
"""
        # make_block_result内でlog_hook_executionが自動呼び出しされる
        result = make_block_result(HOOK_NAME, block_message)
        print(json.dumps(result))
        sys.exit(2)

    except Exception as e:
        print(f"[{HOOK_NAME}] Hook error: {e}", file=sys.stderr)
        result = make_approve_result(HOOK_NAME, f"Hook error: {e}")
        log_hook_execution(HOOK_NAME, "approve", f"Hook error: {e}")
        print(json.dumps(result))
        sys.exit(0)


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/issue_scope_check.ts
#!/usr/bin/env bun
/**
 * Issue編集時のスコープ確認を強制する。
 *
 * Why:
 *   1つのIssueに異なるタスクを追加すると追跡性が低下する。
 *   1Issue1タスクの原則を強制することで、Issueの管理性を向上させる。
 *
 * What:
 *   - gh issue edit --bodyコマンドを検出
 *   - チェックボックスのみの変更は許可（進捗更新のため）
 *   - 内容追加時はスコープ確認を強制しブロック
 *   - SKIP_ISSUE_SCOPE_CHECK環境変数でバイパス可能
 *
 * Remarks:
 *   - ブロック型フック（内容追加時はブロック）
 *   - PreToolUse:Bashで発火（gh issue editコマンド）
 *   - issue-multi-problem-check.pyはIssue作成時のみ対象（責務分離）
 *   - forkセッションではSKIP環境変数を許可しない
 *   - Python版: issue_scope_check.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#2423: チェックボックス更新を許可する機能を追加
 *   - silenvx/dekita#2431: SKIP環境変数サポートと拒否メッセージの改善
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { makeApproveResult, makeBlockResult } from "../lib/results";
import { isForkSession, parseHookInput } from "../lib/session";
import { extractInlineSkipEnv, isSkipEnvEnabled } from "../lib/strings";

const HOOK_NAME = "issue-scope-check";
const SKIP_ENV_NAME = "SKIP_ISSUE_SCOPE_CHECK";

/**
 * Extract Issue number from command.
 */
export function extractIssueNumber(command: string): string | null {
  // gh issue edit 123 --body "..." or gh issue edit #123 -b "..."
  const match = command.match(/gh\s+issue\s+edit\s+#?(\d+)/);
  if (match) {
    return match[1];
  }
  return null;
}

/**
 * Get current Issue body via GitHub API.
 */
function getCurrentIssueBody(issueNumber: string): string | null {
  try {
    const result = execSync(`gh issue view ${issueNumber} --json body --jq ".body"`, {
      encoding: "utf-8",
      timeout: 10000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    return result.trim();
  } catch {
    // gh CLI not installed or timeout
    return null;
  }
}

/**
 * Extract --body option value from command.
 */
export function extractBodyFromCommand(command: string): string | null {
  // --body "$(cat <<'EOF' ... EOF)" pattern (check first)
  const heredocMatch = command.match(
    /--body\s+"\$\(\s*cat\s+<<['"]?EOF['"]?\s*\n(?<body>.*?)\nEOF(?:\)"|"\)|\)|"|\s|$)/s,
  );
  if (heredocMatch?.groups?.body) {
    return heredocMatch.groups.body;
  }

  // --body "value" or --body 'value' pattern
  const quoteMatch = command.match(/--body\s+(['"])(?<body>.*?)\1/s);
  if (quoteMatch?.groups?.body) {
    return quoteMatch.groups.body;
  }

  return null;
}

/**
 * Check if change is checkbox status change only.
 */
export function isCheckboxOnlyChange(oldBody: string, newBody: string): boolean {
  // None or empty body is not "checkbox only change"
  if (!oldBody || !newBody) {
    return false;
  }

  const oldLines = oldBody.split("\n");
  const newLines = newBody.split("\n");

  // Different line count means content added/removed
  if (oldLines.length !== newLines.length) {
    return false;
  }

  // Markdown list markers: -, *, +
  const checkboxPattern = /^(\s*[-*+]\s*)\[([ xX])\](.*)$/;

  for (let i = 0; i < oldLines.length; i++) {
    if (oldLines[i] === newLines[i]) {
      continue;
    }

    const oldMatch = oldLines[i].match(checkboxPattern);
    const newMatch = newLines[i].match(checkboxPattern);

    if (oldMatch && newMatch) {
      // Same prefix and content, only check status differs → OK
      if (oldMatch[1] === newMatch[1] && oldMatch[3] === newMatch[3]) {
        continue;
      }
    }

    // Non-checkbox change found
    return false;
  }

  return true;
}

async function main(): Promise<void> {
  let sessionId: string | undefined;
  try {
    const data = await parseHookInput();
    sessionId = data.session_id;
    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Detect gh issue edit command
    if (!command.includes("gh issue edit")) {
      process.exit(0);
    }

    // Only when changing content with --body option
    if (!command.includes("--body")) {
      process.exit(0);
    }

    // Issue #2458: Fork session check (before SKIP check)
    const source = (data.source as string) || "";
    const transcriptPath = data.transcript_path as string | undefined;
    const isFork = isForkSession(sessionId ?? "", source, transcriptPath);

    // Issue #2431: SKIP env check (export and inline)
    // Issue #2458: Don't allow SKIP in fork session
    const skipRequested =
      isSkipEnvEnabled(process.env[SKIP_ENV_NAME]) ||
      isSkipEnvEnabled(extractInlineSkipEnv(command, SKIP_ENV_NAME));

    if (skipRequested) {
      if (isFork) {
        // Don't allow SKIP in fork session
        await logHookExecution(
          HOOK_NAME,
          "block",
          `fork-session: ${SKIP_ENV_NAME} not allowed`,
          undefined,
          { sessionId },
        );
        const result = makeBlockResult(
          HOOK_NAME,
          `[issue-scope-check] 🚫 forkセッションではSKIP不可

forkセッションでは${SKIP_ENV_NAME}は使用できません。
forkセッションは別タスクとして扱うべきです。

【対応方法】
新しいIssueを作成してください:
gh issue create --title "..." --body "..."`,
        );
        console.log(JSON.stringify(result));
        process.exit(0);
      }

      // Allow SKIP in normal session
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `${SKIP_ENV_NAME}=1: スコープ確認をスキップ`,
        undefined,
        { sessionId },
      );
      const result = makeApproveResult(HOOK_NAME, `${SKIP_ENV_NAME}=1`);
      console.log(JSON.stringify(result));
      process.exit(0);
    }

    // Issue #2423: Allow checkbox status change only
    const issueNumber = extractIssueNumber(command);
    if (issueNumber) {
      const currentBody = getCurrentIssueBody(issueNumber);
      const newBody = extractBodyFromCommand(command);

      if (currentBody && newBody && isCheckboxOnlyChange(currentBody, newBody)) {
        // Checkbox update only, allow
        const result = makeApproveResult(HOOK_NAME, "checkbox status change only");
        await logHookExecution(HOOK_NAME, "approve", "checkbox status change only", undefined, {
          sessionId,
        });
        console.log(JSON.stringify(result));
        process.exit(0);
      }

      // Log why checkbox check was skipped
      if (!currentBody) {
        await logHookExecution(HOOK_NAME, "skip", "Failed to get current issue body", undefined, {
          sessionId,
        });
      }
      if (!newBody) {
        await logHookExecution(
          HOOK_NAME,
          "skip",
          "Failed to extract new body from command",
          undefined,
          { sessionId },
        );
      }
    }

    // Force scope confirmation (block)
    const issueNumForMsg = issueNumber || "<Issue番号>";
    const blockMessage = `🚫 Issue編集時のスコープ確認

Issueに内容を追加する前に確認が必要です:
- 追加しようとしている内容は、元のIssueと同じタスクですか？
- 異なるタスクであれば、別のIssueとして作成すべきです
- 1 Issue = 1 タスク の原則を守ってください

【対応方法】
1. 同じタスクの場合: ユーザーに確認してから編集を続行
2. 異なるタスクの場合: gh issue create --title "..." --body "..." で新規作成

【スキップ方法】（ユーザー確認済みの場合）
\`\`\`
SKIP_ISSUE_SCOPE_CHECK=1 gh issue edit ${issueNumForMsg} --body "..."
\`\`\`

【補足】
- チェックボックスのステータス変更のみの場合は自動許可されます
- 行数が変わる変更（セクション追加など）はブロックされます`;

    const result = makeBlockResult(HOOK_NAME, blockMessage);
    console.log(JSON.stringify(result));
    process.exit(2);
  } catch (error) {
    console.error(`[${HOOK_NAME}] Hook error: ${formatError(error)}`);
    const result = makeApproveResult(HOOK_NAME, `Hook error: ${formatError(error)}`);
    await logHookExecution(HOOK_NAME, "approve", `Hook error: ${formatError(error)}`, undefined, {
      sessionId,
    });
    console.log(JSON.stringify(result));
    process.exit(0);
  }
}

if (import.meta.main) {
  main().catch((e) => {
    console.error(`[${HOOK_NAME}] Fatal error: ${formatError(e)}`);
    console.log(JSON.stringify({}));
  });
}
